resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -542,4 +728,114 @@ private void addTrace(Throwable cause) {
+                return;
+            }
+
+            // Only include servers that serve the correct domain.
+            if (qName.length() <  r.name().length()) {
+                return;
+            }
+
+            String rName = r.name().toLowerCase(Locale.US);
+
+            int dots = 0;
+            for (int a = rName.length() - 1, b = qName.length() - 1; a >= 0; a--, b--) {
+                char c = rName.charAt(a);
+                if (qName.charAt(b) != c) {
+                    return;
+                }
+                if (c == '.') {
+                    dots++;
+                }
+            }
+
+            if (head != null && head.dots > dots) {
+                // We already have a closer match so ignore this one, no need to parse the dmainName etc.
+                return;
+            }
+
+            final ByteBuf recordContent = ((ByteBufHolder) r).content();
+            final String domainName = decodeDomainName(recordContent);
+            if (domainName == null) {
+                // Could not be parsed, ignore.
+                return;
+            }
+
+            // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
+            // we drop servers that have a smaller dots count.
+            if (head == null || head.dots < dots) {
+                count = 1;
+                head = new ServerName(dots, rName, domainName);
+            } else {
+                if (head.dots == dots) {
+                    ServerName serverName = head;
+                    while (serverName.next != null) {
+                        serverName = serverName.next;
+                    }
+                    serverName.next = new ServerName(dots, rName, domainName);
+                    count++;
+                }
+            }
+        }
+
+        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
+        // one node less.
+        ServerName remove(String name) {
+            ServerName serverName = head;
+
+            while (serverName != null) {
+                if (!serverName.removed && serverName.name.equalsIgnoreCase(name)) {
+                    serverName.removed = true;
+                    return serverName;
+                }
+                serverName = serverName.next;
+            }
+            return null;
+        }
+
+        int size() {
+            return count;
+        }
+
+        static final class ServerName {
+            final int dots;
+            final String name;
+            final String delegateName;
+
+            ServerName next;
+            boolean removed;
+
+            ServerName(int dots, String delegateName, String name) {
+                this.dots = dots;
+                this.name = name;
+                this.delegateName = delegateName;
+            }
+
+            boolean isRootServer() {
+                return dots == 1;
+            }
+
+            String delegateName() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename method "delegateName" to prevent any misunderstanding/clash with field "delegateName" defined on line 821. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1845)